snuba/migrations/table_engines.py

```diff
@@ -79,8 +82,26 @@ def __init__(
         super().__init__(order_by, partition_by, sample_by, ttl, settings)
         self.__version_column = version_column
 
-    def _get_engine_type(self, single_node: bool, table_name: str) -> str:
-        if single_node:
+    def _get_engine_type(self, storage_set_key: StorageSetKey, table_name: str) -> str:
+        if get_cluster(storage_set_key).is_single_node():
             return f"ReplacingMergeTree({self.__version_column})"
         else:
             return f"ReplicatedReplacingMergeTree('/clickhouse/tables/{{layer}}-{{shard}})/{table_name}', '{{replica}}', {self.__version_column})"
+
+
+class Distributed(TableEngine):
+    def __init__(self, local_table_name: str, sharding_key: Optional[str]) -> None:
+        self.__local_table_name = local_table_name
+        self.__sharding_key = sharding_key
+
+    def get_sql(self, storage_set_key: StorageSetKey, table_name: str) -> str:
+        cluster = get_cluster(storage_set_key)
+        cluster_name = cluster.get_cluster_name()
+        assert not cluster.is_single_node()
+        assert cluster_name is not None
+        database_name = cluster.get_database()
+        optional_sharding_key = (
+            f", {self.__sharding_key}" if self.__sharding_key else ""
+        )
```

Comment:
I think we have a problem with this approach.
What would the procedure be if we want to merge a two clusters or separating them. Imagine if I had storage_set A assigned to cluster C_1 and storage_set B assigned to cluster C_2. Now I want to move both storage sets on the same cluster.

We need to find a process to allow this. for this use case and document the limitations.

Reply:
For running schema migrations, I think we'll have a tool that allows running the migrations on any specified node regardless of its layout in the cluster. This will be useful for adding brand new nodes to a cluster, bringing up a new cluster or running a specific group migrations on an existing cluster in order to merge them together.

Copying the data over is the next step. We can provide a script to help but (like the schema migrations) it'll be a manual step for the user.

What do you think?

Reply:
If you just changed the settings in step 5 to point at a new (unmigrated) cluster, all the queries to that cluster would immediately fail because the tables are missing. So I think there's no other option but to run all the schema migrations before changing settings. The tool we'd provide would bootstrap a brand new node (or multiple) by running all of the migrations there. Then the data would need to be copied over to the new node(s). Doing this without downtime will be tricky. You probably have to start filling the new table with a new consumer while the old one still runs, and run a script to copy the data to the new cluster. Only once that's done could you safely switch the settings and remove the old consumer and old table.